### PR TITLE
[SPARK-37804][SQL][TEST] Unify v1 and v2 CREATE NAMESPACE tests

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

1. Move `CREATE NAMESPACE` parsing tests to `CreateNamespaceParserSuite`.
2. Put common `CREATE NAMESPACE` tests into one trait `org.apache.spark.sql.execution.command.CreateNamespaceSuiteBase`, and put datasource specific tests to the `v1.CreateNamespaceSuite` and `v2.CreateNamespaceSuite`.

The changes follow the approach of #30287.

### Why are the changes needed?

1. The unification will allow to run common `CREATE NAMESPACE` tests for both DSv1/Hive DSv1 and DSv2
2. We can detect missing features and differences between DSv1 and DSv2 implementations.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing unit tests and new tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 1]
```diff
@@ -1093,93 +1093,6 @@ class DataSourceV2SQLSuite
...
-          val meta =
-            catalog("testcat").asNamespaceCatalog.loadNamespaceMetadata(Array("reservedTest"))
-          assert(meta.get(key) == null || !meta.get(key).contains("foo"),
-            "reserved properties should not have side effects")
-        }
-      }
-    }
-  }
-
-  test("SPARK-37456: Location in CreateNamespace should be qualified") {
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:1165]: This should be covered in `test("namespace with location")` for all catalogs.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    }
+  }
+
+  test("test handling of 'IF NOT EXIST'") {
+    val ns = s"$catalog.$namespace"
+    withNamespace(ns) {
+      sql(s"CREATE NAMESPACE IF NOT EXISTS $ns")
+
+      // The namespace already exists, so this should fail.
+      // TODO: non-Hive catalogs throw NamespaceAlreadyExistsException.
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala:105]: This is a new TODO similar to `// TODO: HiveExternalCatalog throws DatabaseAlreadyExistsException.` Basically, hive catalog throws a different exception (and a slightly different message, requiring `alreadyExistErrorMessage`).

**Comment by imback82** [comment]
cc @cloud-fan @MaxGekk. Happy New Year!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/command/CreateNamespaceSuite.scala [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.apache.spark.sql.execution.command.v1
+
+/**
+ * The class contains tests for the `DESCRIBE NAMESPACE` command to check V1 Hive external
+ * table catalog.
+ */
+class CreateNamespaceSuite extends v1.CreateNamespaceSuiteBase with CommandSuiteBase {
+  override def commandVersion: String = super[CreateNamespaceSuiteBase].commandVersion
+  override def alreadyExistErrorMessage: String = s"$notFoundMsgPrefix $namespace already exists"
```
Comment (imback82) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/command/CreateNamespaceSuite.scala:28]: For Hive, there is no `' '` around the namespace.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/CreateNamespaceSuite.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ *   - V1 In-Memory catalog: `org.apache.spark.sql.execution.command.v1.CreateNamespaceSuite`
+ *   - V1 Hive External catalog:
+ *     `org.apache.spark.sql.hive.execution.command.CreateNamespaceSuite`
+ */
+trait CreateNamespaceSuiteBase extends command.CreateNamespaceSuiteBase
+    with command.TestsV1AndV2Commands {
+  override def namespace: String = "db"
+  override def notFoundMsgPrefix: String = "Database"
+
+  test("Create namespace using default warehouse path") {
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/CreateNamespaceSuite.scala:41]: This is the only difference b/w v1 and v2 catalog.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+ *   - V1 table catalog tests:
+ *     `org.apache.spark.sql.execution.command.v1.CreateNamespaceSuiteBase`
+ *     - V1 In-Memory catalog: `org.apache.spark.sql.execution.command.v1.CreateNamespaceSuite`
+ *     - V1 Hive External catalog:
+*        `org.apache.spark.sql.hive.execution.command.CreateNamespaceSuite`
+ */
+trait CreateNamespaceSuiteBase extends QueryTest with DDLCommandTestUtils {
+  import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+
+  override val command = "Create NAMESPACE"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala:44]:
```suggestion
  override val command = "CREATE NAMESPACE"
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  test("namespace with location") {
+    val ns = s"$catalog.$namespace"
+    withNamespace(ns) {
+      withTempDir { tmpDir =>
+        // The generated temp path is not qualified.
+        val path = tmpDir.getCanonicalPath
+        assert(!path.startsWith("file:/"))
+
+        val e = intercept[IllegalArgumentException] {
+          sql(s"CREATE NAMESPACE $ns Location ''")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala:71]:
```suggestion
          sql(s"CREATE NAMESPACE $ns LOCATION ''")
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        }
+        assert(e.getMessage.contains("Can not create a Path from an empty string"))
+
+        val uri = new Path(path).toUri
+        sql(s"CREATE NAMESPACE $ns Location '$uri'")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala:76]:
```suggestion
        sql(s"CREATE NAMESPACE $ns LOCATION '$uri'")
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+      // TODO: HiveExternalCatalog throws DatabaseAlreadyExistsException.
+      val e = intercept[AnalysisException] {
+        sql(s"CREATE NAMESPACE $ns")
+      }
+      assert(e.getMessage.contains(alreadyExistErrorMessage))
+    }
+  }
+
+  test("test handling of 'IF NOT EXIST'") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/CreateNamespaceSuiteBase.scala:99]: shall we merge this test case to the one above? It's just adding `sql(s"CREATE NAMESPACE IF NOT EXISTS $ns")` to the end of the above test case

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/command/CreateNamespaceSuite.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/command/CreateNamespaceSuite.scala:28]: is this exception thrown by hive? or by Spark?

Comment (imback82) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/command/CreateNamespaceSuite.scala:28]: This is thrown by Hive: `org.apache.hadoop.hive.metastore.api.AlreadyExistsException: Database db already exists` and wrapped to AnalysisException here: https://github.com/apache/spark/blob/9c6f3fbdcfddc78fdca85565ba7b905e26129362/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L111-L112

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**